R/factanal.R

```diff
@@ -0,0 +1,168 @@
+tidy.fa_exploratory <- function(x, type="loadings", n_sample=NULL, pretty.name=FALSE, ...) {
+  if (type == "screeplot") {
+    eigen_res <- eigen(x$correlation, only.values = TRUE) # Cattell's scree plot is eigenvalues of correlation/covariance matrix.
+    res <- tibble::tibble(factor=1:length(eigen_res$values), eigenvalue=eigen_res$values)
+  }
+  else if (type == "variances") {
+    res <- tibble::as_tibble(t(x$Vaccounted)) %>% dplyr::mutate(Factor=as.factor(1:n()), `% Variance`=100*`Proportion Var`, `Cummulated % Variance`=100*`Cumulative Var`)
+  }
+  else if (type == "loadings") {
+    res <- broom:::tidy.factanal(x) # TODO: This just happens to work. Revisit.
+    res <- res %>% tidyr::pivot_longer(cols=c(starts_with("MR"), "uniqueness"), names_to="factor", values_to="value")
+    res <- res %>% dplyr::mutate(factor = case_when(factor=="uniqueness"~"Uniqueness", TRUE~stringr::str_replace(factor,"^MR","Factor ")))
+    res <- res %>% dplyr::mutate(factor = forcats::fct_inorder(factor)) # fct_inorder is to make order on chart right, e.g. Factor 2 before Factor 10
+  }
+  else if (type == "biplot") {
+    scores_df <- broom:::augment.factanal(x)
+    scores_df <- scores_df %>% select(-.rownames) # augment.factanal seems to always return row names in .rownames column.
+    loadings_df <- broom:::tidy.factanal(x)
+
+    if (is.null(n_sample)) { # set default of 5000 for biplot case.
+      n_sample = 5000
+    }
+    # sum of number of loading rows times 2 (because it is line between 2 points) and number of score rows should fit in n_sample.
+    score_n_sample <- n_sample - nrow(loadings_df)*2
+
+    # table of observations. bind original data so that color can be used later.
+    res <- x$df
+
+    # Adjust types of other columns that can be used for color or label.
+    orig_cols <- colnames(res)
+    for (orig_col in orig_cols) {
+      if (!is.numeric(res[[orig_col]])) {
+        if (!is.logical(res[[orig_col]])) {
+          # make categorical columns into factor with NA level, so that legend will show NA.
+          # if we leave them as real NA, legend for NA would not be shown on biplot chart,
+          # since we supress it not to show NAs from the lines for measures.
+          res[[orig_col]] <- forcats::fct_explicit_na(as.factor(res[[orig_col]]), na_level="(NA)")
+        }
+        else {
+          # make logical columns into factor with NA level, so that legend will show NA.
+          res[[orig_col]] <- forcats::fct_explicit_na(factor(res[[orig_col]], levels = c("TRUE","FALSE")), na_level="(NA)")
+        }
+      }
+    }
+
+    res <- res %>% dplyr::bind_cols(scores_df)
+    res <- res %>% sample_rows(score_n_sample)
+
+    # calculate scale ratio for displaying loadings on the same chart as scores.
+    loadings_matrix <- as.matrix(loadings_df %>% dplyr::select(MR1, MR2))
+    max_abs_loading <- max(abs(loadings_matrix))
+    max_abs_score <- max(abs(c(res$MR1, res$MR2)))
+    scale_ratio <- max_abs_score/max_abs_loading
```

Comment:
Add 'divide by 0' if it happens.

Reply:
max_abs_loading becoming 0 would be unlikely. If it would happen, I would expect some other error beforehand.